lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -1004,6 +1029,149 @@ struct BroadcastOpConversion
+          acc = umax(acc, cur);
+        } else {
+          acc = smax(acc, cur);
+        }
+        break;
+      case RedOp::MIN:
+        if (type.isUnsignedInteger()) {
+          acc = umin(acc, cur);
+        } else {
+          acc = smin(acc, cur);
+        }
+        break;
+      case RedOp::FADD:
+        acc = fadd(acc.getType(), acc, cur);
+        break;
+      case RedOp::FMAX:
+        acc = fmax(acc, cur);
+        break;
+      case RedOp::FMIN:
+        acc = fmin(acc, cur);
+        break;
+      case RedOp::XOR:
+        acc = xor_(acc, cur);
+        break;
+      default:
+        llvm::report_fatal_error("Unsupported reduce op");
+      }
+    };
+
+    auto srcTy = op.operand().getType().template dyn_cast<RankedTensorType>();
+    auto srcLayout = srcTy.getEncoding().dyn_cast<BlockedEncodingAttr>();
+    auto srcShape = srcTy.getShape();
+
+    auto llvmElemTy = getTypeConverter()->convertType(srcTy.getElementType());
+    auto elemPtrTy = LLVM::LLVMPointerType::get(llvmElemTy, 3);
+    Value smemBase = getSharedMemoryBase(loc, rewriter, op.getOperation());
+    smemBase = bitcast(elemPtrTy, smemBase);
+
+    std::vector<unsigned> shmShape;
+    for (auto d : srcShape)
+      shmShape.push_back(d);
+    shmShape[axis] =
+        srcLayout.getThreadsPerWarp()[axis] * srcLayout.getWarpsPerCTA()[axis];
+
+    unsigned srcElems = getElemsPerThread(srcLayout, srcShape);
+    auto srcIndices = emitIndices(loc, rewriter, srcLayout, srcShape);
+    auto srcValues =
+        getElementsFromStruct(loc, adaptor.operand(), srcElems, rewriter);
+
+    SmallVector<SmallVector<unsigned>> offset =
+        emitOffsetForBlockedLayout(srcLayout, srcShape);
+
+    std::map<SmallVector<unsigned>, Value> accs;
+    std::map<SmallVector<unsigned>, SmallVector<Value>> indices;
+
+    for (int i = 0; i < srcElems; ++i) {
+      SmallVector<unsigned> key = offset[i];
+      key[axis] = 0;
+      bool is_first = accs.find(key) == accs.end();
+      do_acc(accs[key], srcValues[i], is_first);
+      if (is_first) {
+        indices[key] = srcIndices[i];
+      }
+    }
+
+    Value zero = i32_val(0);
+
+    for (auto it : accs) {
+      const SmallVector<unsigned> &key = it.first;
+      Value acc = it.second;
+      SmallVector<Value> writeIdx = indices[key];
+
+      writeIdx[axis] =
+          udiv(writeIdx[axis], i32_val(srcLayout.getSizePerThread()[axis]));
+      Value writeOffset = linearize(rewriter, loc, writeIdx, shmShape);
+      Value writePtr = gep(elemPtrTy, smemBase, writeOffset);
+      store(acc, writePtr);
+
+      SmallVector<Value> readIdx(writeIdx.size(), zero);
+      for (int N = shmShape[axis] / 2; N > 0; N >>= 1) {
+        readIdx[axis] = i32_val(N);
+        Value readMask = icmp_slt(writeIdx[axis], i32_val(N));
+        Value readOffset =
+            select(readMask, linearize(rewriter, loc, readIdx, shmShape), zero);
+        Value readPtr = gep(elemPtrTy, writePtr, readOffset);
+        barrier;
```

Comment:
A little confused this line at a glance, it reads like putting a variable this line, not like triggering some method/logic. Could we make it `barrier();` here, although it is a macro,  `#define barrier() ...` is fine.

We try to make the macros [there](https://github.com/openai/triton/pull/774/files#diff-8a9a5a7f4a025fb1299af29d190d5626bd9000406d3ea47c49679272d3d6abe9R89) used in a way similar to function calls, e.g. `Value _i32 = int_val(32);`, so `barrier();` is more consistent to existing macros there.

Reply:
done